File: with-asm/asmAss.cpp
// asmAss.cpp

#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include "gad.h"

void Gad::asmAss(MyRT* rt,char* xto,char* xfrom) {
  rt->to("# asmAss "),rt->to(xfrom),rt->to(","),rt->to(xto),rt->to("\n");
  char from[512]; char to[512];
  bool literal = false;
  DType dt = typeOfLiteral(xfrom);
  long dl = 0L;
  switch(dt) {
  case REAL:
    dl = valReal(xfrom);
    sprintf(from,"$%ld",dl);
    rt->to("# movq "); rt->to(xfrom); rt->to(",%rsi\n");
    rt->to("  movq "); rt->to(from); rt->to(",%rsi\n");
    break;
  case NUM:
    dl = valNum(xfrom);
    sprintf(from,"$%ld",dl);
    rt->to("# movq "); rt->to(xfrom); rt->to(",%rsi\n");
    rt->to("  movq "); rt->to(from); rt->to(",%rsi\n");  
    break;
  case UNDEF:
    if(memcmp(xfrom,"gad_",4)!=0) sprintf(from,"%s.%s",rt->curProc,xfrom);
    else strcpy(from,xfrom);
    rt->to("  lea "),rt->to(from),rt->to("(%rip),%rsi\n");
    rt->to("  movq (%rsi),%rsi\n");
    break;
  default: break;
  };
  if(memcmp("gad_",xto,4)!=0) sprintf(to,"%s.%s",rt->curProc,xto);
  else strcpy(to,xto);
  rt->to("  lea "); rt->to(to); rt->to("(%rip),%rdi\n");
  rt->to("  movq %rsi,(%rdi)\n");
}

File: gad-go/asm_ass.go

// asm_ass.go

package main

import (
  "fmt"
  "strings"
)


func AsmAss(xto string ,xfrom string) {
  Wr("# asmAss ", xfrom,",", xto, "\n");
  var from string; var to string;
  var dt = TypeOfLiteral(xfrom);
  switch(dt) {
  case DTYPE_REAL: {
    from = ValReal(xfrom);
    Wr("# movq ", xfrom, ",%rsi\n",
       "  movq $", from, ",%rsi\n");
  }
  case DTYPE_NUM: {
    from = ValNum(xfrom);
    Wr("# movq ", xfrom, ",%rsi\n",
       "  movq $", from, ",%rsi\n"); 
  }
  case DTYPE_UNDEF: {
    if ! strings.HasPrefix(xfrom,"gad_") { from = fmt.Sprintf("%s.%s",CurProc,xfrom);
    } else { from = xfrom; };
    Wr("  lea ", from, "(%rip),%rsi\n",
       "  movq (%rsi),%rsi\n");
  }
  };
  if ! strings.HasPrefix(xto,"gad_") {  to = fmt.Sprintf("%s.%s",CurProc,xto);
  } else { to = xto; };
  Wr("  lea ", to, "(%rip),%rdi\n",
     "  movq %rsi,(%rdi)\n");
}

